src/main/java/io/reactivex/internal/operators/parallel/ParallelJoin.java

```diff
@@ -0,0 +1,368 @@
+                return;
+            }
+
+            drainLoop();
+        }
+
+        void drainLoop() {
+            int missed = 1;
+
+            JoinInnerSubscriber<T>[] s = this.subscribers;
+            int n = s.length;
+            Subscriber<? super T> a = this.actual;
+
+            for (;;) {
+
+                long r = requested.get();
+                long e = 0;
+
+                middle:
+                while (e != r) {
+                    if (cancelled) {
+                        cleanup();
+                        return;
+                    }
+
+                    Throwable ex = error.get();
+                    if (ex != null) {
+                        cleanup();
+                        a.onError(ex);
+                        return;
+                    }
+
+                    boolean d = done.get() == 0;
+
+                    boolean empty = true;
+
+                    for (int i = 0; i < n; i++) {
+                        JoinInnerSubscriber<T> inner = s[i];
+
+                        SimplePlainQueue<T> q = inner.queue;
+                        if (q != null) {
+                            T v = q.poll();
+
+                            if (v != null) {
+                                empty = false;
+                                a.onNext(v);
+                                inner.requestOne();
+                                if (++e == r) {
+                                    break middle;
+                                }
+                            }
+                        }
+                    }
+
+                    if (d && empty) {
+                        a.onComplete();
+                        return;
+                    }
+
+                    if (empty) {
+                        break;
+                    }
+                }
+
+                if (e == r) {
+                    if (cancelled) {
+                        cleanup();
+                        return;
+                    }
+
+                    Throwable ex = error.get();
+                    if (ex != null) {
+                        cleanup();
+                        a.onError(ex);
+                        return;
+                    }
+
+                    boolean d = done.get() == 0;
+
+                    boolean empty = true;
+
+                    for (int i = 0; i < n; i++) {
+                        JoinInnerSubscriber<T> inner = s[i];
+
+                        SimpleQueue<T> q = inner.queue;
+                        if (q != null && !q.isEmpty()) {
+                            empty = false;
+                            break;
+                        }
+                    }
+
+                    if (d && empty) {
+                        a.onComplete();
+                        return;
+                    }
+                }
+
+                if (e != 0 && r != Long.MAX_VALUE) {
+                    requested.addAndGet(-e);
+                }
+
+                int w = get();
```

Comment:
I'm not going to apply them eagerly because it is another local variable/register to worry about when there are lots of other locals in deeper/user code.